plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -145,6 +145,105 @@ func NoDiskConflict(pod *api.Pod, existingPods []*api.Pod, node string) (bool, e
 	return true, nil
 }
 
+type MaxEBSVolumeCountChecker struct {
+	maxVolumes uint
+	pvInfo     PersistentVolumeInfo
+	pvcInfo    PersistentVolumeClaimInfo
+}
+
+// MaxEBSVolumeCountPredicate evaluates whether a pod can fit based on the number of AWS ElasticBlockStorage
+// volumes it requests, and those that are already present.  The maximum number is configurable, although
+// Amazon reccomends that only 40 volumes be mounted, including the system root volume.
+//
+// The predicate looks for both EBS volumes used directly, as well as PVC volumes that are backed by EBS volumes,
+// counts the number of unique volumes, and rejects the new pod if it would place the total count over the maximum.
+func NewMaxEBSVolumeCountPredicate(maxVolumes uint, pvInfo PersistentVolumeInfo, pvcInfo PersistentVolumeClaimInfo) algorithm.FitPredicate {
+	c := &MaxEBSVolumeCountChecker{
+		maxVolumes: maxVolumes,
+		pvInfo:     pvInfo,
+		pvcInfo:    pvcInfo,
+	}
+
+	return c.predicate
+}
+
+func (c *MaxEBSVolumeCountChecker) filterEBSVolumes(volumes []api.Volume, namespace string, ebsVolumes map[string]bool) error {
+	for _, vol := range volumes {
+		if vol.AWSElasticBlockStore != nil {
+			ebsVolumes[vol.AWSElasticBlockStore.VolumeID] = true
+		} else if vol.PersistentVolumeClaim != nil {
+			pvcName := vol.PersistentVolumeClaim.ClaimName
+			if pvcName == "" {
+				return fmt.Errorf("PersistentVolumeClaim had no name: %q", pvcName)
```

Comment:
A pod cannot use the persistentVolumeClaim volume without a ClaimName.  It would fail validation.  It is certainly an error here if there is no claimName.

As I wrote elsewhere in this thread, you need to use claim's pvc.Spec.VolumeName to get the backing PV and inspect its PersistentVolumeSource for a non-nil EBS volume (or GCE, if this PR covers that volume, too).